assets/js/modules/adsense/components/module/ModuleTopEarningPagesWidget/index.js

```diff
@@ -70,15 +71,46 @@ function ModuleTopEarningPagesWidget( {
 				limit: 10,
 			};
 
+			const report = select( MODULES_ANALYTICS ).getReport( reportArgs );
+
+			let pageTitles;
+
+			let hasLoadedPageTitles = false;
+			if ( undefined !== report ) {
+				const pagePaths = [];
+				( report?.[ 0 ]?.data?.rows || [] ).forEach(
+					( { dimensions } ) => {
+						if (
+							dimensions.length &&
+							! pagePaths.includes( dimensions[ 0 ] )
+						) {
+							pagePaths.push( dimensions[ 0 ] );
+						}
+					}
+				);
+				pageTitles = select( MODULES_ANALYTICS ).getPageTitles( {
+					startDate,
+					endDate,
+					pagePaths,
+				} );
+				hasLoadedPageTitles = undefined !== pageTitles;
+			}
+
+			const hasLoaded =
+				hasLoadedPageTitles &&
+				select( MODULES_ANALYTICS ).hasFinishedResolution(
+					'getReport',
+					[ reportArgs ]
+				);
```

Comment:
@eugene-manuilov @ivankruchkoff I'm concerned about the code complexity this implementation is introducing here throughout all these widgets, which is something I was aiming to specifically avoid when we scoped out the IB. I don't think we should require all widgets that require page titles to add all this complicated logic within the `useSelect`.

Could we abstract this somehow so that the complexity goes into a helper function, or maybe even directly into `getPageTitles`? How about this idea:
* Change `getPageTitles` to the following signature `getPageTitles( report, options )` where it would simply receive the `report` (or `undefined`) and the `options` (i.e. report arguments) from where it could get information like the `startDate` and `endDate`.
* If `report` is `undefined`, return `undefined` for the page titles as well.
* If `report` is set/loaded, extract the URLs (we don't need to pass the index within `dimensions` for it, see my other comment) from the `report` and the `startDate` and `endDate` from `options` and run the extra request.

That should simplify the implementation within each widget drastically, something like:
```js
const report = select( MODULES_ANALYTICS ).getReport( reportArgs );
const pageTitles = select( MODULES_ANALYTICS ).getPageTitles( report, reportArgs );
```